features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/net/ResourceRequestBodyFactory.kt

```diff
@@ -0,0 +1,147 @@
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.sessionreplay.internal.net
+
+import com.datadog.android.api.InternalLogger
+import com.datadog.android.api.storage.RawBatchEvent
+import com.datadog.android.sessionreplay.internal.exception.InvalidPayloadFormatException
+import com.datadog.android.sessionreplay.internal.utils.MiscUtils
+import com.google.gson.JsonObject
+import okhttp3.MediaType.Companion.toMediaTypeOrNull
+import okhttp3.MultipartBody
+import okhttp3.RequestBody
+import okhttp3.RequestBody.Companion.toRequestBody
+import kotlin.jvm.Throws
+
+internal class ResourceRequestBodyFactory(
+    private val internalLogger: InternalLogger
+) {
+
+    @Throws(InvalidPayloadFormatException::class)
+    internal fun create(
+        resources: List<RawBatchEvent>
+    ): RequestBody {
+        val deserializedResources: List<ResourceEvent> = resources.mapNotNull {
+            val resourceMetadata = MiscUtils.safeDeserializeToJsonObject(internalLogger, it.metadata)
+
+            if (resourceMetadata != null) {
+                val applicationId = MiscUtils.safeGetStringFromJsonObject(
+                    internalLogger,
+                    resourceMetadata,
+                    APPLICATION_ID_KEY
+                )
+                val filename = MiscUtils.safeGetStringFromJsonObject(
+                    internalLogger,
+                    resourceMetadata,
+                    FILENAME_KEY
+                )
+
+                if (applicationId != null && filename != null) {
+                    ResourceEvent(
+                        applicationId = applicationId,
+                        identifier = filename,
+                        it.data
+                    )
+                } else {
+                    null
+                }
+            } else {
+                null
+            }
+        }
+
+        if (deserializedResources.isEmpty()) {
+            @Suppress("ThrowingInternalException")
+            throw InvalidPayloadFormatException(NO_RESOURCES_TO_SEND_ERROR)
+        }
+
+        val applicationId = getApplicationId(deserializedResources)
+
+        @Suppress("UnsafeThirdPartyFunctionCall") // Handled up in the caller chain
+        val builder = MultipartBody.Builder().setType(MultipartBody.FORM)
+
+        addResourcesSection(builder, deserializedResources)
+        addApplicationIdSection(builder, applicationId)
+        @Suppress("UnsafeThirdPartyFunctionCall") // Handled up in the caller chain
+        return builder.build()
+    }
+
+    private fun getApplicationId(resources: List<ResourceEvent>): String {
+        val applicationIds = resources.map {
+            it.applicationId
+        }
+
+        var selectedApplicationId = applicationIds[0]
+
+        if (applicationIds.size > 1) {
+            // more than one applicationId, so take the one from the largest group
```

Comment:
> [!NOTE]
> What's the reason we use this vs the first or last application Id?

Reply:
My concern is that if we use the first or last applicationId we potentially end up discarding a majority of resources in some cases - an extreme example would be 9 resources with one applicationId followed by one resource with another. We would lose 9 and only send 1.

Reply:
Changed to take the last applicationId in the list